Generate non-written to benchmarking inputs from existing memory instead of new allocation

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122643
* __->__ #121999
* #124031
* #122825
* #123229
* #122642
* #124030

In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Generate non-written to benchmarking inputs from existing memory instead of new allocation

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -3880,18 +3881,53 @@ def __del__(self):
-                src_code = self.codegen_template(
-                    template_node, epilogue_nodes, only_gen_src_code=True
+            # TODO - handle epilogue inputs as well
+            non_out_arg_indices = list(
```
This part of code in unnecessary since you overwrite `non_out_arg_indices` below?

eellison replied:
Oh yea, this was from an old version. will delete

shunting314 reviewed (commented):
So this PR is like we 'steal' storage from weights/activations for benchmarking inputs if the inputs are not mutated:) ?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -170,6 +172,99 @@ def load_triton():
+
+        # We dont want to use the same storage for multiple example inputs,
+        # would distort timing.
+        stor_data_ptr = mod_tensor.untyped_storage().data_ptr()
```
storage_offset() is the difference

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Generate non-written to benchmarking inputs from existing memory instead of new allocation"


In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

<img width="669" alt="image" src="https://github.com/pytorch/pytorch/assets/11477974/836de381-d156-4c46-a70b-bf6ca1799b43">

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
